Add steps to configure a PGP key for agent upgrade

kilfoyle opened the pull request:
This adds instructions for air-gapped environments on how to set `server.pgp.upstream_url` so that Elastic Agents can access the PGP/GPG key required for upgrade.

See [docs preview page](https://ingest-docs_bk_984.docs-preview.app.elstc.co/guide/en/fleet/master/air-gapped.html#air-gapped-pgp-fleet).

Please let me know if I've missed anything, especially in the "Note that" section under the steps.

Closes: #980 
Rel: https://github.com/elastic/sdh-beats/issues/4496

kilfoyle commented:
@lucabelluccini  When you have a chance, please let me know if this [new section](https://ingest-docs_bk_984.docs-preview.app.elstc.co/guide/en/fleet/master/air-gapped.html#air-gapped-pgp-fleet) looks okay. Thanks!

kilfoyle commented:
@lucabelluccini  trying again. Please have a look if you have a chance. Thanks!

kilfoyle commented:
@michel-laterman , @michalpristas , @lucabelluccini 
Can any of you please review and confirm that these added [Configure Elastic Agents to download a PGP/GPG key from Fleet Server](https://ingest-docs_bk_984.docs-preview.app.elstc.co/guide/en/fleet/master/air-gapped.html#air-gapped-pgp-fleet) steps are correct?

lucabelluccini commented:
I still think a developer should review this.

kilfoyle commented:
Thanks @lucabelluccini 

> I still think a developer should review this.

@michel-laterman or @michalpristas  Would either of you mind taking a look at this? Please let me know if you're not the right people to ask.

michel-laterman commented:
I think it looks good, do we want to add anything to explain how to use the fleet-server endpoint as an alternative when distributing gpg keys (issue https://github.com/elastic/fleet-server/issues/2887)?

kilfoyle commented:
> do we want to add anything to explain how to use the fleet-server endpoint as an alternative when distributing gpg keys (issue https://github.com/elastic/fleet-server/issues/2887)?

@michel-laterman I've add the steps below, but I'm way out of my depth here so please let me know whatever needs fixing up.

---

![Screenshot 2024-07-29 at 3 46 04 PM](https://github.com/user-attachments/assets/76942a0e-63f3-4ebe-a28b-781bcaa4de1e)

michel-laterman commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,48 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+
+In an air-gapped environment, the agent won't be able to download the remote key and therefore cannot be upgraded. For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.
+
+Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from a custom URL specified in {fleet-server}:
```
I think my reading comprehension on my 1st pass wasn't great.

Just to be clear fleet-server serves PGP keys at `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`
It will serve the key in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` by default, if there is no `default.pgp` key in the dir it will look upstream to retrieve a key; `server.pgp.upstream_url`  can be used to configure to a custom URL for fleet-server to use.

IIRC agents will try to get the upstream key, then the fleet-server endpoint (@michalpristas please confirm), they do not use the url we add to fleet-server config

kilfoyle replied:
Thanks @michel-laterman! That's really clear. Here's my attempt to fix things up:

---

![Screenshot 2024-07-30 at 12 37 29 PM](https://github.com/user-attachments/assets/7d541046-db25-495c-9326-95dd30e9e3ee)

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,43 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+
+Starting from version 8.9.0, when {agent} tries to perform an upgrade, it first verifies the binary signature with the key bundled in the agent. This process has a backup mechanism that will use the key coming from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` instead of the one it already has.
+
+In an air-gapped environment, the agent won't be able to download the remote key and therefore cannot be upgraded. For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.
```
```suggestion
In an air-gapped environment, an {agent} which doesn't have access PGP/GPG key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` would fail to be upgraded.

For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.
```

kilfoyle replied:
Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,43 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+In an air-gapped environment, the agent won't be able to download the remote key and therefore cannot be upgraded. For versions 8.9.0 to 8.10.3, you can resolve this problem following the steps described in the associated link:https://www.elastic.co/guide/en/fleet/8.9/release-notes-8.9.0.html#known-issues-8.9.0[known issue] documentation.
+
+
+Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from {fleet-server}.
```
```suggestion
Starting in version 8.10.4, {agent} will:
- verify the binary signature with the key bundled in the agent - if it doesn't pass
- download the PGP/GPG key from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` and verify it - if it doesn't pass
- download the PGP/GPG key from {fleet-server} and verify it
```

kilfoyle replied:
Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,43 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+Starting in version 8.10.4, you can resolve this problem by configuring {agents} to download the PGP/GPG key from {fleet-server}.
+
+By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
+The key is served through the endpoint `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
```
```suggestion
The key is served through the Fleet Server endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`.
```

kilfoyle replied:
Thanks! Added in [f698c5e](https://github.com/elastic/ingest-docs/pull/984/commits/f698c5e0a0d5281f279d95865d0c3dd62019841e)

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,43 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+
+By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
+The key is served through the endpoint `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
+Accessing the endpoint requires a valid API key and the endpoint is rate limited.
```
```suggestion
Accessing the endpoint requires a valid API key and the endpoint is rate limited.
```

I am not sure we're protecting the key via API key - can we confirm?

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,43 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+The key is served through the endpoint `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
+Accessing the endpoint requires a valid API key and the endpoint is rate limited.
+
+If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.
```
```suggestion
If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.
```

Few points to clarify:

- Is the `default.pgp` bundled with Fleet Server?

If it's bundled, air-gapped users are good to go with it.

- Is the `default.pgp` downloaded lazily from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` by default?

If it is the case, then air-gapped users can grant access to it (HTTP Proxy or open the firewall).

- Is the `default.php` downloaded lazily from the URL provided via `server.pgp.upstream_url` setting?

If it is the case, then air-gapped users can specify a URL so that it will be downloaded from there.

michel-laterman replied:
It's a lazy download from `server.pgp.upstream_url`

michel-laterman replied:
To be very clear: 
> * Is the `default.pgp` bundled with Fleet Server?

No, on install this file does not exist; but a user can add it manually write the file on the instance if there is 0 internet access in order to distribute it to the agents

> * Is the `default.pgp` downloaded lazily from `https://artifacts.elastic.co/GPG-KEY-elastic-agent` by default?

Yes, the default value of `server.pgp.upstream_url` is `https://artifacts.elastic.co/GPG-KEY-elastic-agent`; if `default.pgp` is not on-disk when the key is requested, fleet-server will download the file from the upstream URL, save it to disk and serve it

> * Is the `default.php` downloaded lazily from the URL provided via `server.pgp.upstream_url` setting?

yes, see above 😄

kilfoyle replied:
@michel-laterman I'm sorry to bring this one up again (I lost track of it and am doing some housekeeping now). 

Is there anything I need to clarify in the text above, or do you think this is okay to merge as is?

michel-laterman replied:
I don't think so

lucabelluccini commented:
Hello @kilfoyle - I've left some comments. We might need clarifications by the developers.
I've not taken the time to setup an environment to try it out yet.

kilfoyle commented:
Thanks @lucabelluccini I've added in your suggestions. @michel-laterman  when you have time I'd appreciate if you can help with Luca's questions.

I've updated the screen capture [here](https://github.com/elastic/ingest-docs/pull/984#discussion_r1697266899) to show the latest version.

shainaraskas commented:
Does this still need a review, or should it be closed?

kilfoyle commented:
@shainaraskas thanks, this can stay open and I plan to address comments.

kilfoyle commented:
@lucabelluccini  I need to merge any remaining open docs PRs by end of day on Wednesday. Are there any remaining changes that you'd like for this PR? I'd suggest that we merge it as is, and if we need to refine it we can do that later. What do you think?

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,49 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+Note that:
+
+ * `server.pgp.upstream_url` may be specified as an `http` endpoint (instead of `https`).
+ * For an `https` endpoint, the CA must be trusted by {fleet-server}.
```
Is there a setting where one can provide the CA to trust the TLS Certificate of the Fleet server?

kilfoyle replied:
@michel-laterman Can you help with Luca's question?

michel-laterman replied:
I believe that it would be the standard CA options

lucabelluccini replied:
This is Fleet Server integration connecting to `server.pgp.upstream_url`. 
Would it use the `--certificate-authorities` (the global one for EA)?

kilfoyle replied:
@michel-laterman Tagging you just in case you haven't seen Luca's question. Thanks!

michel-laterman replied:
Yes, it should

kilfoyle replied:
Thanks! I've changed the text to:

>For an `https` endpoint, the CA for {fleet-server} to connect to `server.pgp.upstream_url` must be trusted by {fleet-server} using the `--certificate-authorities` setting that is used globally for {agent}.

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,49 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+
+[source,yaml]
+----
+server.pgp.upstream_url: <my-web-server-url>
```
It must be the full url or just the host?
E.g. `http://my-web:8080/default.pgp` or `http://my-web:8080/`?

kilfoyle replied:
@michel-laterman Can you help with Luca's question?

kilfoyle replied:
Updated to:

```
In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` 
with the full URL where the PGP/GPG key can be accessed. For example:
[source,yaml]
----
server.pgp.upstream_url: <http://my-web-server:8080/default.pgp>
```

lucabelluccini commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,49 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+By default, {fleet-server} serves {agents} with the key located in `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp`.
+The key is served through the {fleet-server} endpoint `GET /api/agents/upgrades/{major}.{minor}.{patch}/pgp-public-key`.
+
+If there isn't a `default.pgp` key in the `FLEETSERVER_BINARY_DIR/elastic-agent-upgrade-keys/default.pgp` directory, {fleet-server} instead will attempt to retrieve a PGP/GPG key from the URL that you can specify with the `server.pgp.upstream_url` setting.
```
Can we add a statement to tell users can download the file manually instead of waiting for Fleet Server to lazily download the file? This might be useful for Fleet Servers which do not have access to the public URL.

kilfoyle replied:
Thanks! I think this is addressed by your comment below, and I've committed it.

kilfoyle commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,51 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+. Select a policy for the agents that you want to upgrade.
+. On the policy page, in the **Actions** menu for the {fleet-server} integration, select **Edit integration**.
+. In the {fleet-server} settings section expand **Change defaults** and **Advanced options**.
+. In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` with the URL where the PGP/GPG key can be accessed. For example:
+
+[source,yaml]
+----
+server.pgp.upstream_url: <my-web-server-url>
```
```suggestion
. In the **Custom fleet-server configurations** field, add the setting `server.pgp.upstream_url` with the full URL where the PGP/GPG key can be accessed. For example:

[source,yaml]
----
server.pgp.upstream_url: <http://my-web-server:8080/default.pgp>
```

kilfoyle commented on docs/en/ingest-management/fleet/air-gapped.asciidoc:
```diff
@@ -32,6 +32,51 @@ Set the following property in {kib} to enable air-gapped mode in {fleet}. This a
+Note that:
+
+ * `server.pgp.upstream_url` may be specified as an `http` endpoint (instead of `https`).
+ * For an `https` endpoint, the CA must be trusted by {fleet-server}.
```
```suggestion
 * For an `https` endpoint, the CA for {fleet-server} to connect to `server.pgp.upstream_url` must be trusted by {fleet-server} using the `--certificate-authorities` setting that is used globally for {agent}.
```

karenzone reviewed (changes requested):
LGTM

kilfoyle commented:
I'm going to merge this since the docs freeze is tonight and any unmerged work will be lost.
